src/System.Text.Json/tests/Serialization/Array.ReadTests.cs

```diff
@@ -201,6 +201,13 @@ public static void ReadPrimitiveArrayFail()
             Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<int[,]>(Encoding.UTF8.GetBytes(@"[[1,2],[3,4]]")));
         }
 
+        [Fact]
+        public static void ReadArrayDeclaredAsObjectFail()
+        {
+            // JSON type is Dictionary<string, List<SimpleTestClass>>
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<string, SimpleTestClass>>(Encoding.UTF8.GetBytes(@"{""MyList"":[{""MyString"":""name""}]}")));
```

Comment:
Can you add some more test cases in here (: https://github.com/dotnet/corefx/blob/bc78f93494ea732a5852ce46aded6ecfc5135731/src/System.Text.Json/tests/Serialization/DictionaryTests.cs#L394

Looks like a test case with the JSON payload like this exists for a class containing a dictionary, but not for deserializing directly into a dictionary (specifically a dictionary of `<string, Class>`):
https://github.com/dotnet/corefx/blob/bc78f93494ea732a5852ce46aded6ecfc5135731/src/System.Text.Json/tests/Serialization/DictionaryTests.cs#L392

Reply:
Although that test case looks the same, `ClassWithDictionary`'s value type is `int[]`, so, the start array token is expected, whereas my scenario is that `JsonSerializer` should not be expecting an array at when the start array token is read. Also, since `int` is a value type, it goes down a different code path here:

https://github.com/dotnet/corefx/blob/2c06259cf2400704e550a74bf7b044159c2fcce5/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs#L99-L110

As far as I can tell, `HandleObjectAsValue` seems to correctly handle any error condition I can throw at it. However, `HandleStartArray` incorrectly handles when the expected type is an object, not an array.